http/src/v0/pin.rs

```diff
@@ -1,107 +1,98 @@
+use serde::{Deserialize, Serialize};
+use std::borrow::Cow;
+use std::collections::HashMap;
 use std::convert::TryFrom;
-use warp::{reply, Filter, Rejection, Reply};
+use warp::{Filter, Rejection, Reply};
 
-#[derive(Debug)]
-struct AddRequest {
-    args: Vec<Cid>,
-    recursive: bool,
-    progress: bool,
-    // TODO: timeout, probably with rollback semantics?
-}
-
-#[derive(Serialize)]
-struct AddResponse {
-    #[serde(rename = "Pins")]
-    pins: Vec<StringSerialized<Cid>>,
-    // FIXME: go-ipfs doesn't respond with this
-    //progress: u8,
-}
+mod add;
 
 /// `pin/add` per https://docs.ipfs.io/reference/http/api/#api-v0-pin-add or the
 /// interface-ipfs-http test suite.
 pub fn add<T: IpfsTypes>(
     ipfs: &Ipfs<T>,
 ) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
-    with_ipfs(ipfs).and(add_request()).and_then(add_inner)
-}
-
-async fn add_inner<T: IpfsTypes>(
-    ipfs: Ipfs<T>,
-    request: AddRequest,
-) -> Result<impl Reply, Rejection> {
-    if request.recursive {
-        // FIXME: this is not documented however all tests always pass true or false
-        return Err(crate::v0::support::NotImplemented.into());
-    }
-
-    if request.progress {
-        // FIXME: there doesn't appear to be a test for this
-        return Err(crate::v0::support::NotImplemented.into());
-    }
-
-    let cids: Vec<Cid> = request.args;
-
-    let dispatched_pins = cids
-        .into_iter()
-        .map(|x| async { ipfs.pin_block(&x).await.map(move |_| StringSerialized(x)) });
-
-    let completed = try_join_all(dispatched_pins)
-        .await
-        .map_err(StringError::from)?;
-
-    Ok(reply::json(&AddResponse {
-        pins: completed,
-        //progress: 100,
-    }))
-}
-
-pub fn list<T: IpfsTypes>(
-    ipfs: &Ipfs<T>,
-) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
-    with_ipfs(ipfs).and_then(list_inner)
-}
-
-async fn list_inner<T: IpfsTypes>(_ipfs: Ipfs<T>) -> Result<impl Reply, Rejection> {
-    // interestingly conformance tests call this with `paths=cid&stream=true&arg=cid`
-    // this needs to be a stream of the listing
-    Err::<&'static str, _>(crate::v0::NotImplemented.into())
-}
-
-pub fn rm<T: IpfsTypes>(
-    ipfs: &Ipfs<T>,
-) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
-    with_ipfs(ipfs).and_then(rm_inner)
+    with_ipfs(ipfs)
+        .and(add::add_request())
+        .and_then(add::add_inner)
 }
 
-async fn rm_inner<T: IpfsTypes>(_ipfs: Ipfs<T>) -> Result<impl Reply, Rejection> {
-    Err::<&'static str, _>(crate::v0::NotImplemented.into())
+#[derive(Debug)]
+struct ListRequest {
+    // FIXME: should be Vec<IpfsPath>
+    arg: Vec<Cid>,
+    filter: PinFilter,
+    // FIXME: not sure if this is used
+    quiet: bool,
+    // FIXME copypaste
+    stream: bool,
+    timeout: Option<humantime::Duration>,
 }
 
-impl<'a> TryFrom<&'a str> for AddRequest {
+impl<'a> TryFrom<&'a str> for ListRequest {
```

Comment:
nit: `TryFrom<&'a str>` is probably more idiomatically expressed via `FromStr`

Reply:
Yeah, agreed. I actually wanted to get rid of this altogether, since we could use `serde_urlencoded` (or what ever it's correct name was) for everything else than the repeated `arg=ipfspath` with just:

```
#[derive(serde::Deserialize)]
struct ListRequest {
  #[serde(skip)]
  arg: Vec<IpfsPath>,
  ...other_fields
}
```

This would need to have a custom `warp::query::<ListRequest>()` where the raw query string was received, first serde_urldecoded was used and then the repeated argument was parsed into a Vec. I tried to do this with something like:

```
trait ArgReceiver {
  type Arg: std::str::FromStr;

  /// Receive the multiple arg parameters with the option of further validating the args.
  /// Easiest validation is checking if this is empty or not; ListRequest can use empty as well.
  /// We could trigger errors on unsupported `/ipns/` prefix and so on.
  fn receive_args(&mut self, args: Vec<Self::Arg>) -> Result<(), SomeValidationError>;
}
```

But couldn't see a way out of this tanglement and the required bounds on the filter in this hurry. I don't think there are any stopping issues with this, there's just a bit to think through through and get the error types aligned.

Reply:
I will turn this comment into an issue, lets keep this unresolved until.

Reply:
Done: https://github.com/rs-ipfs/rust-ipfs/issues/352